src/main/java/edu/iastate/ece/sd/sdmay2126/orchestration/JobManager.java

```diff
@@ -0,0 +1,146 @@
+    private final BlockingQueue<Runner> availableRunners;
+    /** Jobs which have been prepared and queued for execution. */
+    private final BlockingQueue<Job> jobQueue;
+
+    // Graceful stop indicator, even if mid-queue
+    private boolean stopped = false;
+
+    /**
+     * Creates a job manager with the default runner count.
+     *
+     * @param runnerInstantiator Provisions runners for the manager.
+     */
+    public JobManager(RunnerInstantiator runnerInstantiator) {
+        this(DEFAULT_RUNNER_COUNT, runnerInstantiator);
+    }
+
+    /**
+     * Creates a job manager with the specified runner count.
+     */
+    public JobManager(byte totalRunnerCount, RunnerInstantiator runnerInstantiator) {
+        this.totalRunnerCount = totalRunnerCount;
+        runners = new LinkedList<>();
+        availableRunners = new LinkedBlockingQueue<>();
+        jobQueue = new LinkedBlockingQueue<>();
+
+        // Initialize the runners
+        for (int i = 0; i < totalRunnerCount; i++) {
+            // Note that this does NOT indicate resource availability; that's the runner's responsibility
+            Runner newRunner = runnerInstantiator.createRunner();
+            runners.add(newRunner);
+
+            // Start the runner's initialization process
+            new Thread(newRunner).start();
+        }
+    }
+
+    /**
+     * Schedules a prepared job for execution by some runner, once one comes available.
+     *
+     * @throws JobManagerStoppedException Thrown if the manager was stopped.
+     */
+    public void scheduleJob(Job job) throws JobManagerStoppedException {
+        if (stopped)
+            throw new JobManagerStoppedException();
+
+        jobQueue.add(job);
+    }
+
+    /**
+     * Attempts to unschedule the specified job. Note that this call depends on the caller beating the runners to the job.
+     */
+    public void unscheduleJob(Job job) {
+        jobQueue.remove(job);
+    }
+
+    /**
+     * Indicates that the specified runner has resource availability to process jobs.
+     *
+     * @throws JobManagerStoppedException Thrown if the manager was stopped.
+     */
+    public void indicateAvailability(Runner runner) throws JobManagerStoppedException {
+        if (stopped)
+            throw new JobManagerStoppedException();
+
+        availableRunners.add(runner);
+    }
+
+    /**
+     * Attempts to cancel the specified runner's resource availability. Note that this depends on the caller beating
+     * some job being queued for processing.
+     */
+    public void cancelAvailability(Runner runner) {
+        availableRunners.remove(runner);
+    }
+
+    /**
+     * Instruct the runner to stop execution gracefully.
+     */
+    public void stopManager() {
+        stopped = true;
+    }
+
+    @Override
+    public void run() {
+        // First we have the execution loop
+        try {
+            while (!stopped) {
+                // TODO: Can these calls be made reactive s.t. we can respond to a
```

Comment:
Makes me think something like [LiveData](https://developer.android.com/topic/libraries/architecture/livedata) for Android.

Ultimately it's just an observer pattern. We could set a callback to "stop all jobs" and call that from `stopManager`. That way we set this stopped var and reactively stop the current queue.

Reply:
We could always bring-in [Rx](https://github.com/ReactiveX/RxJava) but I'm not sure our project really warrants it.

I'm not sure the callback would help here - wouldn't it still be stuck blocking on the queues? I wonder if we would just need to force interrupts (if the thread is waiting/blocked), which wouldn't be too bad. I'm not very familiar with best-practices for asynchronous development in Java.